File: reference/harness/completion.py
"""Deterministic completion assurance without mutable local authority.

S5 does not trust a SessionStart snapshot or local remote-tracking references as
completion authority. At Stop it re-evaluates repository posture, reads the
checked default-branch head directly from GitHub, and treats a clean READY default
branch whose local HEAD equals that authoritative GitHub head as a verified
no-change repository state. That state is not, by itself, proof that the task may
finish without repository changes; callers must obtain requirement-side agreement
before accepting the no-change path. Every other or unverifiable state runs the
normal deterministic completion gate.
"""

from __future__ import annotations

import json
import os
import subprocess
from pathlib import Path
from typing import Any, Sequence
from urllib.parse import quote

from reference.posture.checker import check_repository_posture

ROOT = Path(__file__).resolve().parents[2]


def _run_git(cwd: Path, args: Sequence[str]) -> tuple[str | None, str | None]:
    """Run one bounded local Git query used by completion assurance."""
    try:
        result = subprocess.run(
            ["git", *args],
            cwd=cwd,
            text=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            timeout=5,
            check=False,
        )
    except Exception as exc:
        return None, str(exc)
    if result.returncode != 0:
        return None, (result.stderr or result.stdout or f"exit {result.returncode}").strip()
    return result.stdout.strip(), None


def _github_branch_head(
    cwd: Path, repository: str, branch: str
) -> tuple[str | None, str | None]:
    """Return the checked repository's authoritative GitHub branch-head SHA."""
    endpoint = f"repos/{repository}/git/ref/heads/{quote(branch, safe='')}"
    try:
        result = subprocess.run(
            ["gh", "api", endpoint],
            cwd=cwd,
            text=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            timeout=15,
            check=False,
        )
    except Exception as exc:
        return None, str(exc)
    if result.returncode != 0:
        return None, (result.stderr or result.stdout or f"exit {result.returncode}").strip()
    try:
        value = json.loads(result.stdout)
        sha = value.get("object", {}).get("sha") if isinstance(value, dict) else None
    except Exception as exc:
        return None, f"invalid GitHub branch-ref JSON: {exc}"
    if not isinstance(sha, str) or not sha:
        return None, "GitHub branch-ref response does not contain object.sha"
    return sha, None


def capture_session_start(raw: dict[str, Any]) -> str:
    """Return context without persisting authoritative completion state."""
    del raw
    return "Completion assurance will re-evaluate authoritative repository state at Stop."


def _clean_checked_default_branch(cwd: Path) -> tuple[bool, str]:
    """Return whether current state is a verified no-change repository state.

    The state check requires a freshly evaluated ``READY`` posture, the checked
    default branch, a completely clean worktree including untracked files, and an
    exact local-HEAD match with the branch head read directly from GitHub. It does
    not establish that the task requirement permits completion without repository
    changes. Any missing or weaker evidence returns false and leaves the normal
    completion gate in force.
    """
    try:
        report = check_repository_posture(cwd)
    except Exception as exc:
        return False, f"posture evaluation failed: {exc}"
    if report.state != "READY" or not report.repository or not report.default_branch:
        return False, report.summary()

    branch, error = _run_git(cwd, ["branch", "--show-current"])
    if error or not branch or branch != report.default_branch:
        return False, error or "current branch is not the checked default branch"

    status, error = _run_git(cwd, ["status", "--porcelain=v1", "--untracked-files=all"])
    if error or status:
        return False, error or "worktree is not clean"

    head, error = _run_git(cwd, ["rev-parse", "HEAD"])
    if error or not head:
        return False, error or "HEAD is unavailable"
    remote_head, error = _github_branch_head(cwd, report.repository, report.default_branch)
    if error or not remote_head:
        return False, error or "GitHub default branch head is unavailable"
    if head != remote_head:
        return False, "local default branch differs from the authoritative GitHub default branch"
    return True, f"clean READY default branch {report.default_branch} matches GitHub branch head {remote_head}"


def _run_completion_gate(cwd: Path) -> tuple[bool, str]:
    """Run the deterministic repository delivery gate and fail closed on errors."""
    gate = ROOT / "reference" / "scripts" / "completion_gate.sh"
    try:
        completed = subprocess.run(
            ["bash", str(gate)],
            cwd=cwd,
            env=os.environ.copy(),
            text=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.STDOUT,
            timeout=int(os.environ.get("AGENT_HARNESS_COMPLETION_TIMEOUT", "120")),
            check=False,
        )
    except Exception as exc:
        return False, f"completion gate failed closed: {exc}"
    output = completed.stdout.strip()
    return completed.returncode == 0, output or f"completion gate exited {completed.returncode}"


def completion_check(raw: dict[str, Any]) -> tuple[bool, str]:
    """Require agreement for no-change completion or a passing full delivery gate.

    A verified clean default-branch state proves only that no repository change
    remains. It does not prove that completing without a change satisfies the task
    requirement. Until a caller supplies requirement-side agreement, S5 rejects
    that shortcut with an actionable reason so the vendor adapter can ask the
    request source rather than silently accepting an ambiguous completion.

    This guarantee is limited to repository-delivery state. It does not establish
    that the session produced no external side effects.
    """
    cwd = Path(str(raw.get("cwd") or os.getcwd())).resolve()
    no_change_state, reason = _clean_checked_default_branch(cwd)
    if no_change_state:
        return False, (
            "completion assurance requires request-source agreement before "
            f"accepting a no-change repository completion; {reason}"
        )
    ok, gate_reason = _run_completion_gate(cwd)
    return ok, gate_reason

File: internal/completion/completion.go
// Package completion implements vendor-independent deterministic completion assurance.
// It produces evidence for the agent's semantic completion review; deterministic
// success alone is never interpreted as semantic task completion.
package completion

import (
	"context"
	"fmt"
	"strings"

	"github.com/tomo-chan/agent-harness/internal/repository"
)

type Outcome string

const (
	OutcomeBlocked        Outcome = "blocked"
	OutcomeReviewRequired Outcome = "review_required"
	OutcomeComplete       Outcome = "complete"
)

type Check struct {
	Name   string `json:"name"`
	Status string `json:"status"`
	Detail string `json:"detail"`
}

type Evidence struct {
	Repository       string  `json:"repository"`
	Branch           string  `json:"branch"`
	LocalHeadSHA     string  `json:"local_head_sha"`
	DefaultBranch    string  `json:"default_branch"`
	GitHubDefaultSHA string  `json:"github_default_sha,omitempty"`
	NoDeliveryDelta  bool    `json:"no_delivery_delta"`
	Checks           []Check `json:"checks"`
}

type Result struct {
	Outcome  Outcome  `json:"outcome"`
	Reason   string   `json:"reason"`
	Evidence Evidence `json:"evidence"`
}

type Request struct {
	CWD      string
	FollowUp bool
}

type GitHub interface {
	BranchHead(context.Context, string, string) (string, string, error)
}

// Gate executes repository-specific deterministic delivery checks when the
// no-delivery-delta shortcut cannot be proven.
type Gate interface {
	Check(context.Context, string) (string, error)
}

func add(e *Evidence, name, status, detail string) {
	e.Checks = append(e.Checks, Check{Name: name, Status: status, Detail: detail})
}

func blocked(e Evidence, reason string) Result {
	return Result{Outcome: OutcomeBlocked, Reason: reason, Evidence: e}
}

// Evaluate consumes a fresh Repository Authority report produced at Stop time.
// It does not accept SessionStart snapshots or local remote-tracking refs as authority.
func Evaluate(ctx context.Context, req Request, report repository.Report, git repository.Git, github GitHub, gate Gate) Result {
	e := Evidence{Repository: report.Repository, Branch: report.Branch, LocalHeadSHA: report.HeadSHA, DefaultBranch: report.DefaultBranch}
	if req.CWD == "" || report.State != "READY" || report.Repository == "" || report.Branch == "" || report.HeadSHA == "" || report.DefaultBranch == "" {
		add(&e, "repository_authority", "fail", "fresh READY repository evidence is required at Stop")
		return blocked(e, "deterministic completion assurance cannot establish repository authority")
	}
	add(&e, "repository_authority", "pass", "fresh READY repository evidence supplied")
	if git == nil || github == nil || gate == nil {
		add(&e, "completion_provider", "unknown", "Git, GitHub, and deterministic gate providers are required")
		return blocked(e, "deterministic completion evidence is unavailable")
	}

	branch, err := git.Run(ctx, req.CWD, "rev-parse", "--abbrev-ref", "HEAD")
	if err != nil || branch == "" || branch == "HEAD" {
		add(&e, "current_branch", "unknown", "current branch is unavailable")
		return blocked(e, "deterministic completion evidence is unavailable")
	}
	if branch != report.Branch {
		add(&e, "current_branch", "fail", "branch changed after fresh repository authority evaluation")
		return blocked(e, "current branch no longer matches repository authority")
	}
	add(&e, "current_branch", "pass", branch)

	head, err := git.Run(ctx, req.CWD, "rev-parse", "HEAD")
	if err != nil || head == "" {
		add(&e, "local_head", "unknown", "local HEAD is unavailable")
		return blocked(e, "deterministic completion evidence is unavailable")
	}
	if head != report.HeadSHA {
		add(&e, "local_head", "fail", "HEAD changed after fresh repository authority evaluation")
		return blocked(e, "local HEAD no longer matches repository authority")
	}
	add(&e, "local_head", "pass", head)

	noDelta := false
	if branch == report.DefaultBranch {
		status, statusErr := git.Run(ctx, req.CWD, "status", "--porcelain=v1", "--untracked-files=all")
		if statusErr != nil {
			add(&e, "clean_worktree", "unknown", "worktree status is unavailable")
		} else if status != "" {
			add(&e, "clean_worktree", "fail", "worktree contains tracked or untracked changes")
		} else {
			add(&e, "clean_worktree", "pass", "worktree is clean including untracked files")
			remoteHead, _, remoteErr := github.BranchHead(ctx, report.Repository, report.DefaultBranch)
			if remoteErr != nil || remoteHead == "" {
				add(&e, "github_default_head", "unknown", "GitHub default-branch head is unavailable")
			} else {
				e.GitHubDefaultSHA = remoteHead
				if remoteHead == head {
					add(&e, "github_default_head", "pass", remoteHead)
					noDelta = true
				} else {
					add(&e, "github_default_head", "fail", "local default branch differs from GitHub authority")
				}
			}
		}
	}

	if noDelta {
		e.NoDeliveryDelta = true
		add(&e, "deterministic_gate", "pass", "no delivery delta; clean READY default branch matches GitHub head")
	} else {
		detail, gateErr := gate.Check(ctx, req.CWD)
		if gateErr != nil {
			if strings.TrimSpace(detail) == "" { detail = gateErr.Error() }
			add(&e, "deterministic_gate", "fail", detail)
			return blocked(e, "deterministic completion gate failed")
		}
		add(&e, "deterministic_gate", "pass", detail)
	}

	if !req.FollowUp {
		return Result{
			Outcome: OutcomeReviewRequired,
			Reason: "deterministic completion assurance passed; evaluate task requirements, execution results, unresolved concerns, and new findings before completing",
			Evidence: e,
		}
	}
	return Result{Outcome: OutcomeComplete, Reason: "deterministic completion assurance passed after agent completion review", Evidence: e}
}

// StaticGate is useful for adapters/tests that already executed a deterministic
// gate and need to supply its result without changing S5 semantics.
type StaticGate struct { Detail string; Err error }
func (g StaticGate) Check(context.Context, string) (string, error) {
	if g.Err != nil { return g.Detail, g.Err }
	if strings.TrimSpace(g.Detail) == "" { return "deterministic gate passed", nil }
	return g.Detail, nil
}

var _ = fmt.Sprintf
